project-clone/internal/zip/setup.go

```diff
@@ -0,0 +1,168 @@
+//
+// Copyright (c) 2019-2021 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package zip
+
+import (
+	"archive/zip"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"path"
+	"path/filepath"
+
+	"github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+
+	"github.com/devfile/devworkspace-operator/project-clone/internal"
+)
+
+const (
+	tmpDir = "/tmp/"
+)
+
+// SetupZipProject downloads and extracts a zip-type project to the corresponding clonePath.
+func SetupZipProject(project v1alpha2.Project) error {
+	if project.Zip == nil {
+		return fmt.Errorf("project has no 'zip' source")
+	}
+	url := project.Zip.Location
+	clonePath := internal.GetClonePath(&project)
+	projectPath := path.Join(internal.ProjectsRoot, clonePath)
+	if exists, err := internal.DirExists(projectPath); exists {
+		// Assume project is already set up
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("failed to check path %s: %s", projectPath, err)
+	}
+
+	zipFilePath := path.Join(tmpDir, fmt.Sprintf("%s.zip", clonePath))
+	log.Printf("Downloading project archive from %s", url)
+	err := downloadZip(url, zipFilePath)
+	if err != nil {
+		return fmt.Errorf("failed to download archive: %s", err)
+	}
+
+	log.Printf("Extracting project archive to %s", zipFilePath)
+	err = unzip(zipFilePath, projectPath)
```

Comment:
I wonder if we should not drop the top level folder.
Like
```
cat <<EOF | kubectl apply -f -
kind: DevWorkspace
apiVersion: workspace.devfile.io/v1alpha2
metadata:
  name: projects-testing
spec:
  started: true
  template:
    projects:
      - name: dwo
        zip:
          location: https://github.com/devfile/devworkspace-operator/archive/refs/heads/main.zip
        clonePath: test-project-zip
      - name: dwo-2
        git:
          checkoutFrom:
            revision: 0.6.x
          remotes:
            origin: https://github.com/devfile/devworkspace-operator.git
        clonePath: "test-project-git"

    components:
      - name: tools
        container:
          image: quay.io/amisevsk/project-clone:debug
          command:
            - "tail"
            - "-f"
            - "/dev/null"
EOF
```
^ the project is the same but git will be in `test-project-git` when zip in `test-project-zip/devworkspace-operator`.
I'm not sure about the current Theia behavior
![Screenshot_20210622_160606](https://user-images.githubusercontent.com/5887312/122931118-44923900-d375-11eb-9dbc-531faeb71694.png)
![Screenshot_20210622_161354](https://user-images.githubusercontent.com/5887312/122931123-452acf80-d375-11eb-80d8-c053ef930524.png)

Reply:
I'll work on it :+1:

I was mostly going for compatibility with how Theia currently does it, but there's no real reason for that :)